### ISSUE TITLE
Reproducibility, randomness

### ISSUE THREAD

**Issue by benanne** [Lasagne/Lasagne#6]
For the sake of reproducibility It should be possible to run the same piece of code twice and get the exact same result.

In practice this means that any step involving randomness should be able to reuse the same seeds. The cleanest way to achieve this is to allow for custom RNG objects to be used for anything that involves randomness.

An interesting complication is that randomness occurs both on the numpy side of things (e.g. parameter initialization, the order in which training examples are presented), and on the Theano side (e.g. dropout and other forms of training noise).

So I'm wondering how we should implement this. We could make every class or method involving randomness take an optional `rng` argument, but that might get messy. Maybe it's better if it was possible to control this globally? I haven't really thought about this yet.

**Comment by cancan101** [comment]
Right now for dropout, you are constructing a global `MRG_RandomStreams`:  https://github.com/benanne/Lasagne/blob/e090ad4d9cf650b9f452232e73c1e503b01d9460/lasagne/layers/noise.py#L9. 

This gets reproducibly seeded: https://github.com/Theano/Theano/blob/b320ab29a4dc38b4c0035e330eb7dbaa379249cc/theano/sandbox/rng_mrg.py#L1053.

The init code on the other hand uses numpy and does not appear to be seeded.

**Comment by benanne** [comment]
Yeah, but that rng used for dropout should also be configurable imo. We need to work on that :) The init code also needs to use a shared, configurable rng (or it needs to be possible to pass in a custom rng, which is a little more flexible I suppose, or both).

**Comment by f0k** [comment]
> The init code on the other hand uses numpy and does not appear to be seeded.

But it uses a global rng (`np.random`) just like the noise layers, so you can seed it to get reproducible results (`np.random.seed(...)`).

Still I think being able to pass an rng and srng wouldn't be too bad to have as an alternative option.

**Comment by cancan101** [comment]
One other phenomenon I noticed is that at least using the cuDNN library, there seems to be some non determinism in the results from convolutions.

**Comment by f0k** [comment]
> One other phenomenon I noticed is that at least using the cuDNN library, there seems to be some non determinism in the results from convolutions.

The reference says:
> By design, most of cuDNN API routines from a given version generate the same bit-
wise results at every run when executed on GPUs with the same architecture and
the same number of SMs. However, bit-wise reproducibility is not guaranteed across
versions, as the implementation of a given routine may change. With the current release,
the following routines do not guarantee reproducibility because they use atomic add
operations:
* cudnnConvolutionBackwardFilter
* cudnnConvolutionBackwardData

This is not our responsibility then.

**Comment by cancan101** [comment]
Maybe I am doing something wrong, but using the same computer, I am getting (slightly) different results. When I change theano to use CPU, the results match out.

**Comment by cancan101** [comment]
It looks like there is some discussion here: https://github.com/BVLC/caffe/issues/1429

**Comment by f0k** [comment]
This confirms what I cited from the reference: The backward passes in cuDNN are not deterministic. We cannot do anything about that. If you need reproducibility you could use the gemm-based convolution instead by disabling cudnn: `THEANO_FLAGS=optimizer_excluding=conv_dnn`.

**Comment by cancan101** [comment]
This is the code that I am running:
```python
np.random.seed(42)
filter_shape = (32, 1, 3,3)
input_shape = (32, 1, 200, 200)
W = theano.shared(utils.floatX(init.Uniform()(filter_shape)))
input_img = theano.shared(utils.floatX(init.Uniform()(input_shape)))

a = T.nnet.conv2d(input_img, W, 
                  image_shape=input_shape,
                  filter_shape=filter_shape,
                  border_mode='full')

f = theano.function([], a)
```

and then I repeatedly call:
```
print np.asarray(f()).astype(np.float64).sum()
```

**Comment by mheilman** [comment]
In lieu of refactoring a lot of things to follow a pattern like scikit-learn's use of `random_state` keyword arguments, perhaps it would be better to use a module level `RandomState` object (e.g., `lasagne.rng`) instead of numpy's, which any program might modify.

Just a thought.  I haven't looked through the code in detail to figure out whether that really makes sense.

**Comment by benanne** [comment]
That could work. I wonder if there are any use cases we should consider where people will want fine-grained control over RNGs, i.e. they want to use multiple and assign them to each layer. I can't think of any but there might be.

**Comment by ebenolson** [comment]
I sometimes do this (not in Theano but numpy) when I want to control my test/train split but keep data augmentation random. Might come up if data augmentation layers are added.

I think I also used layerwise RNGs for implementing "deterministic" mode for fractional pooling.

**Comment by benanne** [comment]
The train/test split would never be controlled by Lasagne's RNG though, so that's not something we have to take into account. The second one could be. Why would the RNGs need to be separate for that?

**Comment by ebenolson** [comment]
I'm just spitballing here. They don't necessarily need to be, but it was an easy way to generate a stochastic but reproducible pooling pattern.

In general I'm conflicted about this - it's nice having full control over randomness, but I also think it's good if the obvious thing to try (calling `np.random.seed()`) works to get reproducibility.

**Comment by mheilman** [comment]
Could we have a lasagne.rng that defaults to np.random?  That would keep
the current behavior, but then the user could also set that to any
RandomState object.   Maybe there are some important differences I'm
missing...

**Comment by ebenolson** [comment]
It's a reasonable possibility, but means we need to be careful that every stochastic component uses `lasagne.rng` rather than numpy. It was discussed a bit in https://github.com/Lasagne/Lasagne/pull/273, but at the time we decided there were too few random layers to bother with it.

**Comment by f0k** [comment]
> It's a reasonable possibility, but means we need to be careful that every stochastic component uses `lasagne.rng` rather than numpy.

I don't think this would be bad, or unreasonably hard -- it's certainly easier than any other option (it's also easier to forget, but I assume we'd catch this in reviews). And for now, the only files acessing `np.random` for anything else than doctests or tests are `lasagne/init.py` and `lasagne/layers/noise.py`. Adding `rng = np.random` to `lasagne/__init__.py` and then using `lasagne.rng` instead of `np.random` in those two files would add some flexibility for users without changing the current behavior at all. (This actually means we can just as well do this after the first release.)

**Comment by f0k** [comment]
Was resolved in #343.
